### ISSUE TITLE
ResultConsumedException when using Transaction functions rc1 driver

### ISSUE THREAD

**Issue by human-bean-555** [neo4j/neo4j-dotnet-driver#380]
**Neo4j version:** 3.5.11 Community
**Neo4j Mode**: Single instance  
**Driver version**: .NET driver 4.0.0-rc1
**Operating System:** Windows 10

### Steps to reproduce
1. Start Neo4j locally using Neo4j Desktop
2. Start any graph - I've used the Movie Graph with `user : "neo4j"`, `pwd : "1234"`
3. Run the below:

```fsharp
open Neo4j.Driver

let driver = GraphDatabase.Driver("bolt://localhost:7687", AuthTokens.Basic("neo4j", "1234"))
let session = driver.AsyncSession()
let statement = "MATCH (m) RETURN m"

// Auto-commit
let runAutoCommit = 
    session.RunAsync statement
    |> Async.AwaitTask 
    |> Async.RunSynchronously

// Works fine and consumes the result
let autoCommitResults = 
    runAutoCommit.ToListAsync() 
    |> Async.AwaitTask 
    |> Async.RunSynchronously

// Transaction functions
let readTransaction = 
    session.ReadTransactionAsync(fun s -> s.RunAsync statement)
    |> Async.AwaitTask
    |> Async.RunSynchronously

// Throws with Neo4j.Driver.ResultConsumedException
let readTResults = 
    readTransaction.ToListAsync()
    |> Async.AwaitTask
    |> Async.RunSynchronously

```

### Expected behavior
The second query using `session.ReadTransactionAsync( .. )` will run successfully. 
### Actual behavior
The following exception is thrown when trying to consume the results of the Transaction:
```
System.AggregateException: One or more errors occurred. ---> Neo4j.Driver.ResultConsumedException: Cannot access records on this result any more as the result has already been consumed or the query runner where the result is created has already been closed.
   at Neo4j.Driver.Internal.Result.ConsumableResultCursor.FetchAsync()
   at Neo4j.Driver.ResultCursorExtensions.<ToListAsync>d__2.MoveNext()
```
Calling `ConsumeAsync()` works ok and yields the summary;  also using `driver 4.0.0-alpha01` there is no issue.

**Comment by zhenlineo** [comment]
Hi @pauldorehill,

This is the expected behaviour. The transaction result cannot be accessed after a transaction is closed. When using a transaction function method (such as ReadTransactionAsync), before exiting from the function, the transaction will be automatically closed. Thus you cannot have access to the result outside of transaction function.

In previous versions, we allow the result to be passed out of the transaction function method scope by pulling everything into local buffer. As 4.0 we are introducing reactive support, we leave the control of the record stream fully to callers.

In short, to fix the error, you shall move `readTransaction.ToListAsync()` into your transaction func `fun s -> s.RunAsync statement`.

Cheers,
Zhen

**Comment by human-bean-555** [comment]
Hi @zhenlineo,

Thanks - that does make more far more sense in the context of the 4.0 driver.
